srml/support/procedural/src/storage/impls.rs

```diff
@@ -215,166 +249,189 @@ impl<'a> Impls<'a> {
-					previous: Option<&'a #kty>,
-					next: Option<&'a #kty>,
-				}
+		// make sure to use different prefix for head and elements.
+		let final_head_key = if let Some(instance) = instance {
+			let method_name = syn::Ident::new(&format!("build_head_key_once_for_{}", name.to_string()), proc_macro2::Span::call_site());
+			quote!{ #instance::#method_name(#prefix.as_bytes()) }
+		} else {
+			let final_head_key = format!("head of {}", prefix);
+			quote!{ #final_head_key.as_bytes() }
+		};
 
-				impl #linkage {
-					/// Update linkage when this element is removed.
-					///
-					/// Takes care of updating previous and next elements points
-					/// as well as updates head if the element is first or last.
-					pub fn remove<S: #scrate::GenericStorage>(
-						self,
-						storage: &S,
-					) {
-						let next_key = self.next.as_ref().map(|x| #key_for(x));
-						let prev_key = self.previous.as_ref().map(|x| #key_for(x));
-
-						if let Some(prev_key) = prev_key {
-							// Retrieve previous element and update `next`
-							let mut res = Self::read(storage, &*prev_key)
-								.expect("Linkage is updated in case entry is removed; it always points to existing keys; qed");
-							res.1.next = self.next;
-							storage.put(&*prev_key, &res);
-						} else {
-							// we were first so let's update the head
-							Self::write_head(storage, self.next.as_ref());
-						}
+		// TODO TODO: test this with instance
 
-						if let Some(next_key) = next_key {
-							// Update previous of next element
-							let mut res = Self::read(storage, &*next_key)
-								.expect("Linkage is updated in case entry is removed; it always points to existing keys; qed");
-							res.1.previous = self.previous;
-							storage.put(&*next_key, &res);
-						}
-					}
+		// generator for linked map
+		//
+		// NOTE: we split implementation because of recursion limit
+		let mut impls = quote! {
+
+			#[derive(Default, #scrate::parity_codec_derive::Encode, #scrate::parity_codec_derive::Decode)]
+			struct #linkage<Key> {
+				/// Previous element key in storage (None for the first element)
+				previous: Option<Key>,
+				/// Next element key in storage (None for the last element)
+				next: Option<Key>,
+			}
 
-					/// Read the contained data and it's linkage.
-					pub fn read<S: #scrate::GenericStorage>(storage: &S, key: &[u8]) -> Option<(#value_type, #linkage)> {
-						storage.get(key)
+			impl<#traitinstance: #traittype, #instance #bound_instantiable>  #name<#traitinstance, #instance> {
+				/// Update linkage when this element is removed.
+				///
+				/// Takes care of updating previous and next elements points
+				/// as well as updates head if the element is first or last.
+				fn remove<S: #scrate::GenericStorage>(
+					linkage: #linkage<#kty>,
+					storage: &S,
+				) {
+					use #scrate::storage::generator::StorageMap;
+
+					let next_key = linkage.next.as_ref().map(<Self as #scrate::storage::generator::StorageMap<#kty, #typ>>::key_for);
```

Comment:
Can we avoid repetition of this `<Self as #scrate...StorageMa..>`, it's really terrible. maybe do `let as_map = quote! { .. };` ?